display_list/display_list_complexity_metal.cc

```diff
@@ -0,0 +1,758 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/display_list_complexity_metal.h"
+
+// The numbers and weightings used in this file stem from taking the
+// data from the DisplayListBenchmarks suite run on an iPhone 12 and
+// applying very rough analysis on them to identify the approximate
+// trends.
+//
+// Constants of proportionality and trends were chosen to better match
+// larger numbers rather than smaller ones. This may turn out to be the
+// wrong decision, but the rationale here is that with the smaller numbers,
+// we have:
+//
+// a) More noise
+// b) Less absolute difference. If the constant we've chosen is out by 50%
+//    on a measurement that is 0.001ms, that's less of an issue than if
+//    the measurement is 15ms.
+// c) Smaller numbers affect the caching decision negligibly; the caching
+//    decision is likely to be driven by slower ops rather than faster ones.
+//
+// In some cases, a cost penalty is used to figure out the cost of an
+// attribute such as anti-aliasing or fill style. In some of these, the
+// penalty is proportional to some other value such as the radius of
+// a circle. In these cases, we ensure that the penalty is never smaller
+// than 1.0f.
+//
+// The error bars in measurement are likely quite large and will
+// vary from device to device, so the goal here is not to provide a
+// perfectly accurate representation of a given DisplayList's
+// complexity but rather a very rough estimate that improves upon our
+// previous cache admission policy (op_count > 5).
+//
+// There will likely be future work that will improve upon the figures
+// in here. Notably, we do not take matrices or clips into account yet.
+//
+// The scoring is based around a baseline score of 100 being roughly
+// equivalent to 0.0005ms of time. With a 32-bit unsigned integer, this
+// would set the maximum time estimate associated with the complexity score
+// at about 21 seconds, which far exceeds what we would ever expect a
+// DisplayList to take rasterising.
+
+namespace flutter {
+
+DisplayListMetalComplexityCalculator*
+    DisplayListMetalComplexityCalculator::instance_ = nullptr;
+
+DisplayListComplexityCalculator*
+DisplayListMetalComplexityCalculator::GetInstance() {
+  if (instance_ == nullptr) {
+    instance_ = new DisplayListMetalComplexityCalculator();
+  }
+  return instance_;
+}
+
+DisplayListMetalComplexityCalculator::MetalHelper::MetalHelper()
+    : is_complex_(false) {}
+
+unsigned int
+DisplayListMetalComplexityCalculator::MetalHelper::ComplexityScore() {
+  // We overflowed, return int_max
+  if (is_complex_) {
+    return std::numeric_limits<unsigned int>::max();
+  }
+
+  // Calculate the impact of saveLayer
+  unsigned int save_layer_complexity;
+  if (save_layer_count_ > 200) {
```

Comment:
Pulling these out to named constants with comments describing their effect would be helpful to future maintainers.

Reply:
I think there are so many constants that it's best to just have a comment above each one. I'll move the comment out of `saveLayer` and into this method so that it's described at the place where it's actually used